From 5ffb2d38d4c34f0850986063f93b5b6f1ec1cf43 Mon Sep 17 00:00:00 2001
From: Preetam Singh Ranawat <apranawat@codeaurora.org>
Date: Wed, 27 May 2015 11:14:03 +0530
Subject: [PATCH 3/4] hal: Fix alignement of buffer sent to DSP for
 multichannel clips

 - currently buffer size is Aligned with 96 which is not multiple
   of some channels(5,7).
 - Buffer size must be multiple of (number of channels * bytes per sample).
   For writes to succeed, the buffer must be written at address which is
   multiple of 32.
 - Alignments of (number of channels * bytes per sample)*32 satisfies both
   of the above requirements.

Change-Id: I20de875615141a4a331383a5348abd28b97306f7
---
 hal/msm8974/platform.c | 9 ++++-----
 1 file changed, 4 insertions(+), 5 deletions(-)

diff --git a/hal/msm8974/platform.c b/hal/msm8974/platform.c
index 6a79960..bd844ca 100644
--- a/hal/msm8974/platform.c
+++ b/hal/msm8974/platform.c
@@ -2353,15 +2353,14 @@ uint32_t platform_get_pcm_offload_buffer_size(audio_offload_info_t* info)
                      * info->sample_rate
                      * (bits_per_sample >> 3)
                      * popcount(info->channel_mask))/1000;
-    // To have same PCM samples for all channels, the buffer size requires to
-    // be multiple of (number of channels * bytes per sample)
-    // For writes to succeed, the buffer must be written at address which is multiple of 32
-    // Alignment of 96 satsfies both of the above requirements
-    fragment_size = ALIGN(fragment_size, 96);
     if(fragment_size < MIN_PCM_OFFLOAD_FRAGMENT_SIZE)
         fragment_size = MIN_PCM_OFFLOAD_FRAGMENT_SIZE;
     else if(fragment_size > MAX_PCM_OFFLOAD_FRAGMENT_SIZE)
         fragment_size = MAX_PCM_OFFLOAD_FRAGMENT_SIZE;
+    // To have same PCM samples for all channels, the buffer size requires to
+    // be multiple of (number of channels * bytes per sample)
+    // For writes to succeed, the buffer must be written at address which is multiple of 32
+    fragment_size = ALIGN(fragment_size, ((bits_per_sample >> 3)* popcount(info->channel_mask) * 32));
 
     ALOGI("PCM offload Fragment size to %d bytes", fragment_size);
     return fragment_size;
-- 
2.7.4

